File: app/src/main/java/openfoodfacts/github/scrachx/openfood/utils/AnalyticsEvent.java
package openfoodfacts.github.scrachx.openfood.utils;

public class AnalyticsEvent {
    public String category;
    public String action;
    public String name;
    public Float value;

    private AnalyticsEvent(String category, String action, String name, Float value) {
        this.category = category;
        this.action = action;
        this.name = name;
        this.value = value;
    }

    public void track() {
        AnalyticsService.getInstance().trackEvent(this);
    }

    public static AnalyticsEvent ScannedBarcode(String barcode) {
        return new AnalyticsEvent("scanner", "scanned", barcode, null);
    }

    public static AnalyticsEvent ScannedBarcodeResultExpanded(String barcode) {
        return new AnalyticsEvent("scanner", "result-expanded", barcode, null);
    }

    public static AnalyticsEvent AllergenAlertCreated(String allergenCode) {
        return new AnalyticsEvent("allergen-alerts", "created", allergenCode, null);
    }

    public static AnalyticsEvent ProductCreated(String barcode) {
        return new AnalyticsEvent("products", "created", barcode, null);
    }

    public static AnalyticsEvent ProductEdited(String barcode) {
        return new AnalyticsEvent("products", "edited", barcode, null);
    }

    public static AnalyticsEvent ProductIngredientsPictureEdited(String barcode) {
        return new AnalyticsEvent("products", "edited-ingredients-picture", barcode, null);
    }

    public static AnalyticsEvent UserLogin() {
        return new AnalyticsEvent("user-account", "login", null, null);
    }

    public static AnalyticsEvent UserLogout() {
        return new AnalyticsEvent("user-account", "logout", null, null);
    }

    public static AnalyticsEvent RobotoffLoginPrompt() {
        return new AnalyticsEvent("user-account", "login-prompt", "robotoff", null);
    }

    public static AnalyticsEvent RobotoffLoggedInAfterPrompt() {
        return new AnalyticsEvent("user-account", "logged-in-after-prompt", "robotoff", null);
    }

    public static AnalyticsEvent ShoppingListCreated() {
        return new AnalyticsEvent("shopping-lists", "created", null, null);
    }

    public static AnalyticsEvent ShoppingListExported() {
        return new AnalyticsEvent("shopping-lists", "exported", null, null);
    }

    public static AnalyticsEvent IngredientAnalysisEnabled(String name) {
        return new AnalyticsEvent("ingredient-analysis", "enabled", name, null);
    }

    public static AnalyticsEvent IngredientAnalysisDisabled(String name) {
        return new AnalyticsEvent("ingredient-analysis", "disabled", name, null);
    }

    public static AnalyticsEvent AddProductToComparison(String barcode) {
        return new AnalyticsEvent("products", "compare-add", barcode, null);
    }

    public static AnalyticsEvent CompareProducts(int count) {
        return new AnalyticsEvent("products", "compare-multiple", null, (float) count);
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/analytics/AnalyticsEvent.kt
package openfoodfacts.github.scrachx.openfood.analytics

import openfoodfacts.github.scrachx.openfood.models.Barcode

sealed class AnalyticsEvent(val category: String, val action: String, val name: String?, val value: Float?) {

    object ProductSearch :
        AnalyticsEvent("search", "completed", null, null)

    object ProductSearchStart :
        AnalyticsEvent("search", "started", null, null)

    data class BarcodeDecoder(val success: Boolean = false, val duration: Float = 0f) :
        AnalyticsEvent("scanner", "scanning", success.toString(), duration)

    data class ScannedBarcode(val barcode: String) :
        AnalyticsEvent("scanner", "scanned", barcode, null)

    data class ScannedBarcodeResultExpanded(val barcode: String?) :
        AnalyticsEvent("scanner", "result-expanded", barcode, null)

    data class AllergenAlertCreated(val allergenTag: String) :
        AnalyticsEvent("allergen-alerts", "created", allergenTag, null)

    data class ProductCreated(val barcode: String?) :
        AnalyticsEvent("products", "created", barcode, null)

    data class ProductEdited(val barcode: String?) :
        AnalyticsEvent("products", "edited", barcode, null)

    data class ProductIngredientsPictureEdited(val barcode: String?) :
        AnalyticsEvent("products", "edited-ingredients-picture", barcode, null)

    object UserLogin :
        AnalyticsEvent("user-account", "login", null, null)

    object UserLogout :
        AnalyticsEvent("user-account", "logout", null, null)

    object RobotoffLoginPrompt :
        AnalyticsEvent("user-account", "login-prompt", "robotoff", null)

    object RobotoffLoggedInAfterPrompt :
        AnalyticsEvent("user-account", "logged-in-after-prompt", "robotoff", null)

    object ShoppingListCreated :
        AnalyticsEvent("shopping-lists", "created", null, null)

    data class ShoppingListProductAdded(val barcode: String) :
        AnalyticsEvent("shopping-lists", "add_product", barcode, null)

    data class ShoppingListProductRemoved(val barcode: String) :
        AnalyticsEvent("shopping-lists", "remove_product", barcode, null)

    object ShoppingListShared :
        AnalyticsEvent("shopping-lists", "shared", null, null)

    object ShoppingListExported :
        AnalyticsEvent("shopping-lists", "exported", null, null)

    data class IngredientAnalysisEnabled(val type: String) :
        AnalyticsEvent("ingredient-analysis", "enabled", type, null)

    data class IngredientAnalysisDisabled(val type: String) :
        AnalyticsEvent("ingredient-analysis", "disabled", type, null)

    data class AddProductToComparison(val barcode: Barcode) :
        AnalyticsEvent("products", "compare-add", barcode.raw, null)

    data class CompareProducts(val count: Float) :
        AnalyticsEvent("products", "compare-multiple", null, count)
}
